File: go/problem_86.go
package main

func minParenthesesRemoval(s string) int {
	open, close := 0, 0
	for _, c := range s {
		if c == '(' {
			open++
		} else {
			if open > 0 {
				open--
			} else {
				close++
			}
		}
	}
	return open + close
}

func problem86() {
	println("Problem 86:")
	assertEqual(minParenthesesRemoval("()())()"), 1)
	assertEqual(minParenthesesRemoval(")("), 2)
	assertEqual(minParenthesesRemoval("()()"), 0)
	println("All Tests Passed")
}

File: python/problem_86.py
def min_parens_removal(s: str) -> int:
    open_count = close_count = 0
    for c in s:
        if c == '(':
            open_count += 1
        else:
            if open_count > 0:
                open_count -= 1
            else:
                close_count += 1
    return open_count + close_count


def main():
    print("Problem 86:")
    assert min_parens_removal("()())()") == 1
    assert min_parens_removal(")(")      == 2
    assert min_parens_removal("()()")    == 0
    print("All Tests Passed")
